### DISCUSSION TITLE
Cargo: proposed new rangeStrategy

### DISCUSSION THREAD

**Discussion by sunshowers** [renovatebot/renovate#28280] [Request Help]
### What would you like help with?

Other

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitHub, 37.280.0

### Please tell us more about your question or problem

Hi -- we have a Renovate setup going over at Oxide for our Rust and npm dependencies, and we really love it!

I have:

1. Some questions about how to achieve a goal.
2. If my understanding of the current world is correct, a proposal for 1-2 new range strategies to be added to Renovate.
3. In addition, a proposal to also change the default from "bump" to one of these new strategies.

## Context

Oxide is a Rust-heavy company, though we also use npm/JavaScript in a few places. I'm going to focus on Rust in this discussion because that's what I'm most familiar with.

We sometimes specify Rust dependencies in this fashion:

```toml
[dependencies]
chrono = "0.4"
syn = "2"
```

And we almost always check in `Cargo.lock` as well.

(Note that for Cargo, "0.4" actually means "~0.4", and "2" or "2.0" means "^2.0.0".)

A thing we've observed is that by default, we don't get Renovate PRs for updates if the full version (x.y.z) isn't specified. So, for example, we currently have [`chrono 0.4.34` in one of our repos](https://github.com/oxidecomputer/omicron/blob/534a940c2bd2074331b51440e43ecf517a4162bc/Cargo.lock#L964-L977), but haven't gotten a PR to update to [version 0.4.37](https://docs.rs/crate/chrono/0.4.37).

I did some testing and it looks like there's an "update-lockfile" strategy that does achieve this goal. But that can't be globally specified because if a full version `x.y.z` is specified in `Cargo.toml`, then that doesn't get bumped with this strategy.

This is easiest to summarize in a table. Below, we assume that version `x.y.z` is being updated to `u.v.w`, where the versions may be compatible or incompatible (e.g. `u` may be the same as `x`, or not.)

| `rangeStrategy`   | spec         | update type  | update `Cargo.lock` | update `Cargo.toml`               |
|-------------------|--------------|--------------|---------------------|-----------------------------------|
| `bump`            | `x` or `x.y` | compatible   | no                  | no                                |
| `bump`            | `x` or `x.y` | incompatible | yes                 | yes, to `u` or `u.v` respectively |
| `bump`            | `x.y.z`      | compatible   | yes                 | yes, to `u.v.w`                   |
| `bump`            | `x.y.z`      | incompatible | yes                 | yes, to `u.v.w`                   |
| `update-lockfile` | `x` or `x.y` | compatible   | yes                 | no                                |
| `update-lockfile` | `x` or `x.y` | incompatible | yes                 | yes, to `u` or `u.v` respectively |
| `update-lockfile` | `x.y.z`      | compatible   | yes                 | no                                |
| `update-lockfile` | `x.y.z`      | incompatible | yes                 | yes, to `u.v.w`                   |

As a workaround, we've come up with a hybrid strategy. This strategy makes it so that if versions are specified as `x` or `x.y`, then we use the `update-lockfile` strategy. Otherwise, we use the bump strategy. We've implemented this via a regex, see [this config](https://github.com/oxidecomputer/renovate-config/blob/f0f0d58e1657b1067467be76c92ba4de8dbff20b/rust/crates.json#L16-L18).

For the hybrid strategy, we have:

| `rangeStrategy`   | spec         | update type  | update `Cargo.lock` | update `Cargo.toml`               |
|-------------------|--------------|--------------|---------------------|-----------------------------------|
| hybrid            | `x` or `x.y` | compatible   | yes                 | no                                |
| hybrid            | `x` or `x.y` | incompatible | yes                 | yes, to `u` or `u.v` respectively |
| hybrid            | `x.y.z`      | compatible   | yes                 | yes, to `u.v.w`                   |
| hybrid            | `x.y.z`      | incompatible | yes                 | yes, to `u.v.w`                   |

And this strategy works well enough. But ideally, we actually want to always change the version in `Cargo.toml`. In other words, we would like to have this strategy:

| `rangeStrategy`   | spec         | update type  | update `Cargo.lock` | update `Cargo.toml`               |
|-------------------|--------------|--------------|---------------------|-----------------------------------|
| desired           | `x` or `x.y` | compatible   | yes                 | yes, to `u.v.w`                   |
| desired           | `x` or `x.y` | incompatible | yes                 | yes, to `u.v.w`                   |
| desired           | `x.y.z`      | compatible   | yes                 | yes, to `u.v.w`                   |
| desired           | `x.y.z`      | incompatible | yes                 | yes, to `u.v.w`                   |

To the best of my knowledge, this desired strategy does not exist -- I couldn't find a set of flags which would enforce the strategy.

## Questions

Based on the above discussions, I have a few questions:

1. Is there an easier way to achieve the "hybrid" strategy mentioned above? I'm glad it is possible to configure this, but based on my reading of the Renovate source code there didn't seem to be a way to achieve that in a simpler fashion.
2. Does the "desired" strategy above make sense? I couldn't see how to achieve this at all based on my read of the documentation and Renovate source code. (It's quite possible I've missed something, though!)
3. If the answer to 1 or 2 is "no", would you consider adding these as additional `rangeStrategy` options?
4. I really feel like the hybrid strategy above should be the default at least for Cargo. It's pretty surprising that some lockfile updates just get ignored. What do you think?

I hope the above description is detailed enough. Always happy to provide more details!

## Other notes

A repo where I ran some tests is at https://github.com/sunshowers/renovate-test. This test repo uses [the hybrid strategy](https://github.com/sunshowers/renovate-test/blob/main/renovate.json), and you can see the results [in the list of PRs](https://github.com/sunshowers/renovate-test/pulls).

The periodic [`lockFileMaintenance` job](https://docs.renovatebot.com/configuration-options/#lockfilemaintenance) sadly hasn't been working for us because it's all-or-nothing: there are some newer versions of deps that are compatible and some that aren't (particularly git deps), so updating them one-by-one is necessary.

Again, I want to say that Renovate is incredibly cool and we overall love our setup. I hope this request can help improve things for all Renovate users.

### Logs (if relevant)

<details><summary>Logs</summary>

Here's an example log for the repo linked above:

https://github.com/sunshowers/renovate-test/blob/main/example.log

I ran this with

```
docker run -v ./config.json:/config.json -e RENOVATE_TOKEN=***** -e RENOVATE_CONFIG_FILE=/config.json -e LOG_LEVEL=debug -e RENOVATE_PR_HOURLY_LIMIT=10 --rm renovate/renovate
```

</details>

**Comment by rarkins** [comment]
Thanks for your thoughtful analysis. Trying to break this into some smaller pieces, we can go back to other parts later if necessary.

If rangeStrategy=update-lockfile and there's an in-range update (e.g. from 1.2.3 to 1.2.4) then you should get a PR. Are you sure it's not because you had 1.2.4 in your lockfile, i.e. you're already updated? If you have 1.2.3 in Cargo.toml/lock and no update for a 1.2.4, please create a reproduction repo so we can treat that as a bug.

Next, I think that rangeStrategy=auto should default to update-lockfile instead of bump for the cargo manager. i.e. update the locked version by default but not keep bumping the version in Cargo.toml

**Reply by sunshowers** [reply]
Thanks for the response!

> If rangeStrategy=update-lockfile and there's an in-range update (e.g. from 1.2.3 to 1.2.4) then you should get a PR. Are you sure it's not because you had 1.2.4 in your lockfile, i.e. you're already updated? If you have 1.2.3 in Cargo.toml/lock and no update for a 1.2.4, please create a reproduction repo so we can treat that as a bug.

We do get a PR in this case -- this is captured in these rows in the table above:

| `rangeStrategy`   | spec         | update type  | update `Cargo.lock` | update `Cargo.toml`               |
|-------------------|--------------|--------------|---------------------|-----------------------------------|
| `update-lockfile` | `x` or `x.y` | compatible   | yes                 | no                                |
| `update-lockfile` | `x.y.z`      | compatible   | yes                 | no                                |


Our concern is that we _do_ want the `Cargo.toml` to be bumped when the lockfile gets bumped -- i.e. the `update Cargo.toml` column to be "yes".

> Next, I think that rangeStrategy=auto should default to update-lockfile instead of bump for the cargo manager. i.e. update the locked version by default but not keep bumping the version in Cargo.toml

Ah interesting. I do think that is a more logically sound approach than the current one with `bump`, because it always produces lockfile updates. At Oxide we could retain our hybrid strategy through a config update as well.

Once we've sorted `update-lockfile` out, I'd love your thoughts on the rest of the post. Thanks again!

**Reply by rarkins** [reply]
> Our concern is that we _do_ want the `Cargo.toml` to be bumped when the lockfile gets bumped -- i.e. the `update Cargo.toml` column to be "yes".

That's what `bump` is meant to do, not `update-lockfile`. It means you are continuously narrowing the range to one version, although it's not as locked down as if you put a pinned version in your Cargo.toml

**Reply by sunshowers** [reply]
Right, hence my proposal to add 1-2 new rangeStrategies. :)

**Reply by rarkins** [reply]
Could you explain or rewrite "spec" and "compatible" above? spec is not a term we use in Renovate, and I'm not sure exactly what you mean by compatible. Without this, I do not fully understand your tables.

**Reply by rarkins** [reply]
BTW when I'm reviewing the Cargo versioning code, we may actually want to change the bump functionality. Right now it looks like a value of `5` won't be changed unless `6.0.0` or later comes out. However that `bump` behavior is then the same as `replace`. Instead maybe `bump` should update to the latest version e.g. `5.1.7`

**Reply by sunshowers** [reply]
Sure.

* "spec" is the semver version requirement: the supported versions recorded in `Cargo.toml`. Looks like [npm calls it "ranges"](https://www.npmjs.com/package/semver#ranges) -- I guess Renovate uses the same terminology, hence "rangeStrategy".

  The term is short for "specification", in the sense that you specify a range of supported versions -- my bad, it's a term I use for it but it doesn't seem to be common.

* "compatible" means an update to a version that's compatible with the range. For example, if the version specified is "^5.1", an update to 5.1.5 or 5.2.0.

**Reply by sunshowers** [reply]
> Right now it looks like a value of 5 won't be changed unless 6.0.0 or later comes out. However that bump behavior is then the same as replace. Instead maybe bump should update to the latest version e.g. 5.1.7

Ah yes, you are right that `bump` as implemented for Cargo seems to be exactly the same as `replace` (going by the docs, haven't tested it). If I'm understanding correctly, you're proposing that `bump` starts behaving like the "desired" strategy proposed in the OP. Is that correct?

**Reply by rarkins** [reply]
Yes, I think it seems to match with your "desired".

The current approach literally calls "replace" if there's no ^ or ~.

I am making this change now across the affected versioning schemes (it's not just cargo): https://github.com/renovatebot/renovate/pull/28309

**Reply by sunshowers** [reply]
Wonderful, thanks!

An issue with the current behavior of `bump` is that it will actually not send PRs for dependencies in some cases -- for example, if the range is `^5.2` and an update to `5.2.1` is provided. This is captured in row 1 of the table in the OP:

| `rangeStrategy`   | spec         | update type  | update `Cargo.lock` | update `Cargo.toml`               |
|-------------------|--------------|--------------|---------------------|-----------------------------------|
| `bump`            | `x` or `x.y` | compatible   | no                  | no                                
|

Does #28309 address that as well?

> The current approach literally calls "replace" if there's no ^ or ~.

Note that Cargo's ranges have an implicit `^` or `~` if an explicit comparator isn't specified. I believe Renovate handles this already, but wanted to reiterate this comment from the OP:

> (Note that for Cargo, "0.4" actually means "~0.4", and "2" or "2.0" means "^2.0.0".)

**Reply by rarkins** [reply]
> An issue with the current behavior of `bump` is that it will actually not send PRs for dependencies in some cases -- for example, if the range is `^5.2` and an update to `5.2.1` is provided.

I cannot reproduce that on main branch - it appears to work as you wish it to - so please create a reproduction repository to demonstrate if you're sure of it

**Reply by sunshowers** [reply]
Sure. Here's a reproduction repository that covers all four rows of the `bump` strategy listed in the table above, including all possible variants of each row:

https://github.com/sunshowers/renovate-test-2

The debug logs for the execution are stored [here](https://github.com/sunshowers/renovate-test-2/blob/main/initial-log-20240410.txt). [Here's the `Cargo.toml` file containing descriptions of each row, and outcomes.](https://github.com/sunshowers/renovate-test-2/blob/main/Cargo.toml)

Feel free to test against your own copy of this repo, or adapt any of it to your needs.

**Reply by rarkins** [reply]
Can you guide me to exactly where/how I can find the particular case I was interested in?

**Reply by sunshowers** [reply]
That would be this line: https://github.com/sunshowers/renovate-test-2/blob/c019b7777d3dd616efd763a1281fede6f795f5a1/Cargo.toml#L16

* The range is 1.0 aka ^1.0.
* The version in Cargo.lock is 1.0.1
* There is an update to 1.0.5 available, which is compatible with ^1.0. https://crates.io/crates/debug-ignore
* With the bump strategy, no PR is created.

(Would it help to have a quicker way to communicate? Happy to get on a direct chat if it'll save you time. Not meant to be an imposition, just an offer to help.)

**Reply by sunshowers** [reply]
Hi -- wanted to ask if there was anything else I can help with in terms of understanding the issue. Hopefully the above instructions are clear enough. Thanks!

**Reply by rarkins** [reply]
Addressed in https://github.com/renovatebot/renovate/pull/28632

**Reply by sunshowers** [reply]
Thank you! Let me try that now.

**Reply by sunshowers** [reply]
This is awesome! It looks like `bump` now creates PRs in all cases. To go back and update the table:

| `rangeStrategy` | spec         | update type  | update `Cargo.lock` | update `Cargo.toml`                                      |
|-----------------|--------------|--------------|---------------------|----------------------------------------------------------|
| new `bump`      | `x` or `x.y` | compatible   | **yes** (old: no)   | **yes, to `u.v.w`** (old: no) ([example 1])              |
| new `bump`      | `x` or `x.y` | incompatible | yes                 | yes, **to `u.v.w`** (old: to `u` or `u.v`) ([example 2]) |
| new `bump`      | `x.y.z`      | compatible   | yes                 | yes, to `u.v.w`                                          |
| new `bump`      | `x.y.z`      | incompatible | yes                 | yes, to `u.v.w`                                          |

[example 1]: https://github.com/sunshowers/renovate-test-2/pull/11
[example 2]: https://github.com/sunshowers/renovate-test-2/pull/7

This matches our desired strategy exactly. Thanks!

There's still this outstanding point:

> Next, I think that rangeStrategy=auto should default to update-lockfile instead of bump for the cargo manager. i.e. update the locked version by default but not keep bumping the version in Cargo.toml

Is this something you'd still like to do? I personally prefer `bump` but would understand if the default was `update-lockfile`.

**Reply by rarkins** [reply]
Yes, I still plan to do it. Essentially it means that by default, we assume that if your Cargo.toml has a range (e.g. 1.0.0) and not a pinned version (=1.0.0) then we assume you want to keep that range and not have it bumped/narrowed. It's a bit too opinionated to bump by default because it's not suitable for most libraries.

**Reply by sunshowers** [reply]
All right, happy to close the discussion since our specific problem has been addressed, or leave it open if you'd like. Thanks so much again!

**Comment by kornelski** [comment]
This change has caused unwanted churn for me ;(

`Cargo.toml` specifies *minimum* allowed versions, rather than the exact versions of a package to use, so it doesn't need to be updated as often as `Cargo.lock`.

Updating `Cargo.toml` is more disruptive than just updating `Cargo.lock`, because `Cargo.toml` can be manually edited to add dependencies or toggle optional features, so needless version bumps can cause merge conflicts. Because `Cargo.toml` is a source of truth, these merge conflicts must be resolved manually. OTOH `Cargo.lock` can be easily rebuilt, so it's not a big deal when it conflicts.

I suggest at least don't bump `semver-patch` versions in `Cargo.toml` for versions above 1.0.0, unless in response to security patches.

**Reply by sunshowers** [reply]
@kornelski could you use the `update-lockfile` strategy?

(edit: fixed word order :) )

**Reply by rarkins** [reply]
Yes, please switch from the default rangeStrategy=auto to either rangeStrategy=replace (which would mean no updates at all unless package file needs changing) or rangeStrategy=update-lockfile (which means updates only to the lock file, unless package file also need updating such as for major updates)
